File: frontend/src/lib/services/photo-report.ts
import { api } from "../api-client";
import type { GenerateResponse, ListParams, Paginated, RecordStatus } from "./common";

/** 列表元素(API.md §4.2)。 */
export type PhotoReportListItem = {
  id: string;
  title: string | null;
  inspection_unit: string | null;
  status: RecordStatus;
  created_at: string;
  updated_at: string;
};

/** 报告图片(详情内嵌)。 */
export type PhotoReportImage = {
  id: string;
  uploaded_file_id: string;
  frame_timestamp: number | null;
  caption: string | null;
  detected_address: string | null;
  detected_violation: string | null;
  is_selected: boolean;
  sort_order: number;
  created_at: string;
};

/** 详情响应(API.md §4.2)。 */
export type PhotoReportDetail = {
  id: string;
  title: string | null;
  inspection_unit: string | null;
  inspection_address: string | null;
  violation_summary: string | null;
  status: RecordStatus;
  images: PhotoReportImage[];
  source_task_id: string | null;
  created_at: string;
  updated_at: string;
};

/** PUT 图片更新:按 id 逐项匹配,仅 caption / is_selected / sort_order 可编辑,不涉及增删。 */
export type PhotoReportImageUpdate = {
  id: string;
  caption?: string | null;
  is_selected?: boolean;
  sort_order?: number;
};

/** PUT 可编辑字段,均为可选,未提交字段保持不变。 */
export type PhotoReportUpdate = {
  title?: string | null;
  inspection_unit?: string | null;
  inspection_address?: string | null;
  violation_summary?: string | null;
  status?: RecordStatus;
  images?: PhotoReportImageUpdate[];
};

const base = "/api/photo-report";

export const photoReportService = {
  generate: (form: FormData) => api.post<GenerateResponse>(`${base}/generate`, form),
  list: (params: ListParams = {}, signal?: AbortSignal) =>
    api.get<Paginated<PhotoReportListItem>>(base, {
      query: { page: params.page, page_size: params.page_size, status: params.status },
      signal,
    }),
  get: (id: string, signal?: AbortSignal) =>
    api.get<PhotoReportDetail>(`${base}/${encodeURIComponent(id)}`, { signal }),
  update: (id: string, patch: PhotoReportUpdate) =>
    api.put<PhotoReportDetail>(`${base}/${encodeURIComponent(id)}`, patch),
  /** 下载 Word 文书(文件流);文档未生成返回 409,由 api-client 转为 ApiError。 */
  download: (id: string) =>
    api.get<Blob>(`${base}/${encodeURIComponent(id)}/download`, { responseType: "blob" }),
};

File: backend/app/schemas/photo_report.py
"""Photo report schemas (API.md §4.2)."""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel

RecordStatus = Literal["draft", "processing", "generated", "reviewed", "finalized", "archived", "failed"]


class PhotoReportImageUpdate(BaseModel):
    id: str
    caption: str | None = None
    is_selected: bool | None = None
    sort_order: int | None = None


class PhotoReportImageOut(BaseModel):
    id: str
    uploaded_file_id: str
    frame_timestamp: float | None = None
    caption: str | None = None
    detected_address: str | None = None
    detected_violation: str | None = None
    is_selected: bool
    sort_order: int
    created_at: datetime


class PhotoReportListItem(BaseModel):
    id: str
    title: str | None = None
    inspection_unit: str | None = None
    status: str
    created_at: datetime
    updated_at: datetime


class PhotoReportListResponse(BaseModel):
    items: list[PhotoReportListItem]
    total: int
    page: int
    page_size: int


class PhotoReportOut(BaseModel):
    id: str
    title: str | None = None
    inspection_unit: str | None = None
    inspection_address: str | None = None
    violation_summary: str | None = None
    status: str
    images: list[PhotoReportImageOut] = []
    source_task_id: str | None = None
    created_at: datetime
    updated_at: datetime


class PhotoReportUpdate(BaseModel):
    title: str | None = None
    inspection_unit: str | None = None
    inspection_address: str | None = None
    violation_summary: str | None = None
    status: RecordStatus | None = None
    images: list[PhotoReportImageUpdate] | None = None
